### PR TITLE
Addressing Numpy bugs (IDs: 37701221, 37701269)

### PR DESCRIPTION
**Pull request by osamaesmailmsft**
###### Merge Checklist  

**All** boxes should be checked before merging the PR *(just tick any boxes which don't apply to this PR)*
- [x] The toolchain has been rebuilt successfully (or no changes were made to it)
- [x] The toolchain/worker package manifests are up-to-date
- [x] Any updated packages successfully build (or no packages were changed)
- [x] Packages depending on static components modified in this PR (Golang, `*-static` subpackages, etc.) have had their `Release` tag incremented.
- [x] Package tests (%check section) have been verified with RUN_CHECK=y for existing SPEC files, or added to new SPEC files
- [x] All package sources are available
- [x] cgmanifest files are up-to-date and sorted (`./cgmanifest.json`, `./toolkit/tools/cgmanifest.json`, `./toolkit/scripts/toolchain/cgmanifest.json`, `.github/workflows/cgmanifest.json`)
- [x] LICENSE-MAP files are up-to-date (`./SPECS/LICENSES-AND-NOTICES/data/licenses.json`, `./SPECS/LICENSES-AND-NOTICES/LICENSES-MAP.md`, `./SPECS/LICENSES-AND-NOTICES/LICENSE-EXCEPTIONS.PHOTON`)
- [x] All source files have up-to-date hashes in the `*.signatures.json` files
- [x] `sudo make go-tidy-all` and `sudo make go-test-coverage` pass
- [x] Documentation has been updated to match any changes to the build system
- [x] Ready to merge

---

###### Summary 

What does the PR accomplish, why was it needed?
Addresses the bugs listed above, which indicate issues with numpy's %check and numpy's own tests.
Also needed to update our version of Cython to be >= 0.29.30 since that's what numpy requires.

###### Change Log  

- Updated numpy.spec 's %check section to install `hypothesis`
- Moved openblas from SPECS-EXTENDED to SPECS
- Upgraded Cython from 0.29.26 to 0.29.32

###### Does this affect the toolchain?  

Yes. Cython is updated from 0.29.26 to 0.29.32

###### Test Methodology

Local AMD64 build, running buddy builds at the moment (AMD64: 265121, ARM64: 265122)

### CODE REVIEW

File: SPECS/numpy/generate_source_tarball.sh
Comment (nicogbg) [inline SPECS/numpy/generate_source_tarball.sh:5]: why not using $PKG_VERSION ?

File: SPECS/numpy/generate_source_tarball.sh
Comment (PawelWMS) [inline SPECS/numpy/generate_source_tarball.sh:7]: I'd remove the hard-coded version, so the `-z` checks below make sense. This will make it clearer to the caller how to use this script and we limit the risk of setting inconsistent arguments (for instance setting the source tarball but forgetting to set the package version).

File: SPECS/numpy/numpy.spec
Comment (nicogbg) [inline SPECS/numpy/numpy.spec:17]: I'd recommend to add 1.23 in zip file name so blob store and have several versions of this guy (this will solve issue that can happen if we move to version 1.24.x later, e.g.: 1.24.x for Mariner 3.0 and still on 1.23.x for Mariner 2.0)
You might also want to automate downloading of that zip file in generate_source_tarball.sh

Comment (nicogbg) [reply SPECS/numpy/numpy.spec:17]: e.g.: "Source1:  https://numpy.org/doc/1.23/numpy-html.zip#/numpy-html-1.23.zip"
fyi, RPM build only care about what's after the last '/'

**Review by nicogbg — CHANGES_REQUESTED** [review]

File: SPECS/numpy/generate_source_tarball.sh
Comment (osamaesmailmsft) [reply SPECS/numpy/generate_source_tarball.sh:5]: I was just using the script as given, but am cleaning up the parameters as noted.

File: SPECS/Cython/Cython.spec
Comment (oliviacrain) [inline SPECS/Cython/Cython.spec:8]: License reference: https://spdx.org/licenses/Apache-2.0.html

Since we're moving to SPDX short identifiers for licenses, mind making the following change?

```suggestion
License:        Apache-2.0
```

File: toolkit/Makefile
Comment (oliviacrain) [inline toolkit/Makefile:139]: These changes look unrelated to this PR. Also, GitHub is saying the PR "wants to merge 197 commits into microsoft:main". Can you rebase your changes on top of the latest main branch revision?

**Review by oliviacrain — DISMISSED** [review]

File: toolkit/Makefile
Comment (osamaesmailmsft) [reply toolkit/Makefile:139]: They're not directly related to this PR, but building openblas takes a few hours, so I used this addition to the Makefile to skip that. I left it in this PR because I thought it might be useful for anyone else using dependencies from extended that take a long time to build. Working on the rebasing.

Comment (oliviacrain) [reply toolkit/Makefile:139]: I really like the change! We should separate it out from the packaging changes. One half of this PR is packaging bugfixes, the other half is a toolkit change to improve developer experience. Separate PRs allow us to have a more nuanced and focused conversation about each topic.

Comment (osamaesmailmsft) [reply toolkit/Makefile:139]: Gotcha! I'll get on that then :)

**Review by nicogbg — APPROVED** [review]

**Review by nicogbg — APPROVED** [review]

File: SPECS/numpy/numpy.spec
Comment (PawelWMS) [inline SPECS/numpy/numpy.spec:2]: We usually don't want to use these releases:
```suggestion
```

File: SPECS/numpy/numpy.spec
Comment (PawelWMS) [inline SPECS/numpy/numpy.spec:51]: Not needed - Mariner never provided this package.
```suggestion
```

File: SPECS/numpy/numpy.spec
Comment (PawelWMS) [inline SPECS/numpy/numpy.spec:78]: Also not needed `Obsoletes` - we never had these packages anyway.

A bit of Mariner trivia: what's probably more important, we've "reset" the epoch numbers on almost all of our packages, so `numpy-f2py < 1:1.10.1-3` would be even problematic, if we ever introduced it as a separate package.

```suggestion
Provides:       f2py = %{version}-%{release}
Provides:       numpy-f2py = %{version}-%{release}
```

File: SPECS/numpy/numpy.spec
Comment (PawelWMS) [inline SPECS/numpy/numpy.spec:128]: Nit-pick: this pushd-popd seems a bit excessive for what we're doing.
```suggestion
ln -s f2py3 %{buildroot}%{_bindir}/f2py.numpy
```

File: SPECS/numpy/numpy.spec
Comment (PawelWMS) [inline SPECS/numpy/numpy.spec:154]: We skip the `%check` section anyway, if `with_check` is set to 0, so we can simplify:
```suggestion
export PYTHONPATH=%{buildroot}%{python3_sitearch}
pip install pytest hypothesis typing-extensions
# test_ppc64_ibm_double_double128 is unnecessary now that ppc64le has switched long doubles to IEEE format.
# https://github.com/numpy/numpy/issues/21094
%ifarch %{ix86}
# Weird RuntimeWarnings on i686, similar to https://github.com/numpy/numpy/issues/13173
# Some tests also overflow on 32bit
%global ix86_k and not test_vector_matrix_values and not test_matrix_vector_values and not test_identityless_reduction_huge_array and not (TestKind and test_all)
%endif
python3 runtests.py --no-build -- -ra -k 'not test_ppc64_ibm_double_double128 %{?ix86_k}'
```

File: SPECS/numpy/numpy.spec
Comment (PawelWMS) [inline SPECS/numpy/numpy.spec:145]: We've had some issue with `pytest` breaking our builds, if allowed to install the latest version like this. Please take a look at the version currently downloaded for these tests and "freeze" it here like so:
```suggestion
pip install pytest==[some_version_number] hypothesis typing-extensions
```

File: SPECS/numpy/numpy.spec
Comment (PawelWMS) [inline SPECS/numpy/numpy.spec:152]: Can be dropped - we don't build for these architectures and I don't think we will.
```suggestion
```

File: SPECS/numpy/numpy.spec
Comment (PawelWMS) [inline SPECS/numpy/numpy.spec:203]: We could also mention that we reset the epoch to indicate it was intentional.

File: SPECS/openblas/openblas.spec
Comment (PawelWMS) [inline SPECS/openblas/openblas.spec:18]: For extended -> core moves we must also clean-up the specs a bit, because the bar is lower for the extended specs. Please lint the spec and remove references to distro-specific macros (for instance `rhel` or `fedora`) or things that do not matter for Mariner (like unsupported architectures).

**Review by PawelWMS — CHANGES_REQUESTED** [review]
A few clean-up requests before we merge.

File: SPECS/openblas/openblas.spec
Comment (PawelWMS) [inline SPECS/openblas/openblas.spec:27]: Please move this conditional `BuildRequires` and `Provides` below, next to the other ones.

File: SPECS/openblas/openblas.spec
Comment (PawelWMS) [inline SPECS/openblas/openblas.spec:25]: It looks like we're setting `lapacke` to 1 in both branches of the `if`, so I'd double-check if that's correct and if so, just remove the macro and assume it's always set to 1 to clean-up its uses in the spec.

File: SPECS/openblas/openblas.spec
Comment (PawelWMS) [inline SPECS/openblas/openblas.spec:233]: To be honest, we can also assume that we're always building for a 64-bit architecture and remove the `build64` macro completely. Not a blocker, completely ignorable for this PR, though.

File: SPECS/openblas/openblas.spec
Comment (PawelWMS) [inline SPECS/openblas/openblas.spec:396]: This line change should be reverted. It's one of the remaining issues with out linter: it doesn't differentiate between macros, that are just simple values/strings, and ones that are functions accepting arguments. Surrounding the latter in curly brackets makes them being run with no arguments and thing end up breaking or misconfigured. Having said that, I don't see this macro even being defined anywhere, so it might be worth taking a look at the build logs, but for this review I'd just need to remove the brackets.

```suggestion
%multilib_fix_c_header --file %{_includedir}/openblas/openblas_config.h
```

**Review by PawelWMS — APPROVED** [review]
Left a few short comments but pre-approving to not block the PR once they are addressed.

**Pull request merged**
